apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -168,13 +172,16 @@ pre_process_args(Mod, Func, Args) ->
         false -> Args
     end.
 
-safe_publish(RuleId, Topic, QoS, Flags, Payload) ->
+safe_publish(RuleId, Topic, QoS, Flags, Payload, PubProps) ->
     Msg = #message{
         id = emqx_guid:gen(),
         qos = QoS,
         from = RuleId,
         flags = Flags,
-        headers = #{republish_by => RuleId},
+        headers = #{
+            republish_by => RuleId,
+            properties => pub_props_to_packet(PubProps)
```

Comment:
please implement it locally.
```suggestion
            properties => pub_props_to_user_properties(PubProps)
```

Reply:
sorry, maybe `_user_properties` is not a good suffix since it includes more than just `User-Properties` headers.